Add basic support for Exec auth

johnpoth opened the pull request:
Adds basic support for Exec Auth (browser is not supported). Tested on EKS.  #32 #53
Thanks!

astefanutti commented on lib/auth/gcp.js:
```diff
@@ -22,28 +22,28 @@ class GoogleCloudPlatform {
 
   refresh_token() {
-    let promise, cancel;
+    let promise, cancel = cancel = function (){};
```
Seems like there is an extra `cancel`?

astefanutti commented on lib/config/user.js:
```diff
@@ -119,10 +129,24 @@ class GoogleCloudPlatformAuthProvider {
 }
 
+class ExecAuthProvider {
+  constructor({ 'apiVersion': api_version, args, command, env}) {
```
nit: add extra space before `}`

astefanutti commented on lib/config/user.js:
```diff
@@ -26,13 +26,17 @@ class User {
       this.auth_provider = new AuthProvider(Object.assign({ name: auth_provider.name, config: auth_provider.config }));
     }
+    if (typeof exec !== 'undefined') {
+      // for now let's treat exec like an Auth Provider
```
nit: `Auth Provider` -> `AuthProvider`

johnpoth commented:
Thanks for reviewing! I've addressed the comments and I added another commit to support Digital Ocean. Let me know if you prefer them squashed?

Should we add something to the readme now or wait when we're about to release? I was thinking of a support matrix browser/no browser auth support. Thanks!

astefanutti commented:
> Thanks for reviewing! I've addressed the comments and I added another commit to support Digital Ocean. Let me know if you prefer them squashed?

Better keeping the two commits. 

> Should we add something to the readme now or wait when we're about to release? I was thinking of a support matrix browser/no browser auth support. Thanks!

I'm fine with updating the README now, even if it should theoretically be done post-release. Anyway, I was thinking cutting a release ASAP, with all the awesome stuff you've added. I like the idea of the support matrix. I would suggest to put it into a dedicated section in the README.

johnpoth commented on lib/auth/exec-auth.js:
```diff
@@ -0,0 +1,55 @@
+      promise = Promise.resolve(this.auth_provider.token);
+    }
+    promise = new Promise((resolve, reject) => {
+      const process = execFile(this.auth_provider.command, this.auth_provider.args , {'env' : this.auth_provider.env}, (error, stdout, stderr) => {
```
I don't have a strong opinion here, it would definitely simplify the code and I agree on the drawback

astefanutti replied:
Let's keep it that way for now 👍

johnpoth commented:
> > Thanks for reviewing! I've addressed the comments and I added another commit to support Digital Ocean. Let me know if you prefer them squashed?
> 
> Better keeping the two commits.
> 
> > Should we add something to the readme now or wait when we're about to release? I was thinking of a support matrix browser/no browser auth support. Thanks!
> 
> I'm fine with updating the README now, even if it should theoretically be done post-release. Anyway, I was thinking cutting a release ASAP, with all the awesome stuff you've added. I like the idea of the support matrix. I would suggest to put it into a dedicated section in the README.

Added a note in the README

astefanutti commented on README.adoc:
```diff
@@ -131,6 +131,24 @@ This can be achived with the {uri-kube-apiserver}[Kubernetes API server CLI], e.
 
+== Authentication
+
+We try to support the various authentication strategies supported by https://kubernetes.io/docs/reference/kubectl/overview/[kubectl] in order to provide seamless integration with your local setup. Here are the different authentication strategies we support depending on how you're using Kubebox:
```
kubectl -> `kubectl`

astefanutti commented on README.adoc:
```diff
@@ -131,6 +131,24 @@ This can be achived with the {uri-kube-apiserver}[Kubernetes API server CLI], e.
 
+== Authentication
+
+We try to support the various authentication strategies supported by https://kubernetes.io/docs/reference/kubectl/overview/[kubectl] in order to provide seamless integration with your local setup. Here are the different authentication strategies we support depending on how you're using Kubebox:
```
kubectl -> `kubectl`

astefanutti replied:
Just format kubectl as an inline code block

johnpoth replied:
Not sure how to do that in a hyperlink, tried

https://kubernetes.io/docs/reference/kubectl/overview/[[source shell]`kubectl`]

astefanutti replied:
https://kubernetes.io/docs/reference/kubectl/overview/[`kubectl`] should work

johnpoth replied:
Done, thanks!

astefanutti commented on README.adoc:
```diff
@@ -131,6 +131,24 @@ This can be achived with the {uri-kube-apiserver}[Kubernetes API server CLI], e.
+
+|OpenID Connect | icon:check[] | icon:check[] | icon:check[] | icon:check[] | icon:check[]
+
+|Exec (Amazon EKS, Digital Ocean) | icon:check[] | | | |
```
I would not mention `Exec`, which is rather an implementation details, and add lines for EKS and DO.

astefanutti commented on README.adoc:
```diff
@@ -131,6 +131,24 @@ This can be achived with the {uri-kube-apiserver}[Kubernetes API server CLI], e.
+|===
+||Executable |Server |Kubectl | Docker| Online
+
+|OpenID Connect | icon:check[] | icon:check[] | icon:check[] | icon:check[] | icon:check[]
```
I would add a table note mentioning that custom IDP certificate authority file is not supported in Web versions.

astefanutti replied:
Do you think that's worth adding for the online mode?

johnpoth replied:
For sure! Added a note

astefanutti commented on README.adoc:
```diff
@@ -131,6 +131,24 @@ This can be achived with the {uri-kube-apiserver}[Kubernetes API server CLI], e.
+
+[cols="<,^,^,^,^,^", options="header"]
+|===
+||Executable |Server |Kubectl | Docker| Online
```
nit: fix spaces

astefanutti commented on README.adoc:
```diff
@@ -131,6 +131,24 @@ This can be achived with the {uri-kube-apiserver}[Kubernetes API server CLI], e.
+
+[cols="<,^,^,^,^,^", options="header"]
+|===
+||Executable |Server |Kubectl | Docker| Online
```
For Docker, I wonder whether we could add the binaries to the image to make it working for GKE, EKS and DO?

johnpoth replied:
That's what I was thinking too however I'm not sure what we should do about different versions of GKE, EKS and DO. We may be lucky and they are all backwards compatible.

Also GKE, EKS and DO require some initial configuration (setting up tokens in DO, GKE for external access). We could pass them as ENV vars to docker though and set them up ourselves.

Additionally, EKS uses cmd-path which we would need to change.

astefanutti replied:
Agreed, that seems quite complex and I'm not sure that many people use the Docker client. Let's keep it that way.

astefanutti commented on README.adoc:
```diff
@@ -131,6 +131,24 @@ This can be achived with the {uri-kube-apiserver}[Kubernetes API server CLI], e.
+
+[cols="<,^,^,^,^,^", options="header"]
+|===
+||Executable |Server |Kubectl | Docker| Online
```
I'm not sure whether this is really applicable to the _server_ and _kubectl_ modes, as they use service accounts to authenticate and do not rely on kube config?

johnpoth replied:
You're right! Updated

johnpoth commented:
Updated PR, thanks!

astefanutti commented:
Thanks! I'll try to cut a release ASAP.
